src/platform_linux/shmem.c

```diff
@@ -0,0 +1,1152 @@
+   // Always trace destroy of shared memory segment.
+   platform_default_log("Deallocated SplinterDB shared memory "
+                        "segment at %p, shmid=%d."
+                        " Used=%lu bytes (%s, " FRACTION_FMT(4, 2) " %%)"
+                        ", Free=%lu bytes (%s, " FRACTION_FMT( 4, 2) " %%)"
+                        ".\n",
+                        shmaddr,
+                        shmid,
+                        shm_used_bytes,
+                        size_str(shm_used_bytes),
+                        (FRACTION_ARGS(used_bytes_pct) * 100),
+                        shm_free_bytes,
+                        size_str(shm_free_bytes),
+                        (FRACTION_ARGS(free_bytes_pct) * 100));
+   // clang-format on
+
+   size_t shm_nf_search_skipped = shminfo->shm_nf_search_skipped;
+   size_t shm_nfrees            = shminfo->shm_nfrees;
+   platform_default_log(
+      "SplinterDB shared memory stats:"
+      " nfrees=%lu, nf_search_skipped=%lu (%d %%)"
+      ", Large fragments in-use HWM=%u (found in-use=%d)"
+      ", consumed=%lu bytes (%s)"
+      ".\n",
+      shm_nfrees,
+      shm_nf_search_skipped,
+      (shm_nfrees ? (int)((shm_nf_search_skipped * 100) / shm_nfrees) : 0),
+      frags_inuse_HWM,
+      found_in_use,
+      used_by_large_frags_bytes,
+      size_str(used_by_large_frags_bytes));
+}
+
+/*
+ * -----------------------------------------------------------------------------
+ * platform_shm_alloc() -- Allocate n-bytes from shared memory segment.
+ *
+ * Allocation request is expected to have added-in pad-bytes required for
+ * alignment. As a result, we can assert that the addr-of-next-free-byte is
+ * always aligned to PLATFORM_CACHELINE_SIZE.
+ * -----------------------------------------------------------------------------
+ */
+void *
+platform_shm_alloc(platform_heap_id hid,
+                   const size_t     size,
+                   const char      *objname,
+                   const char      *func,
+                   const char      *file,
+                   const int        lineno)
+{
+   shmem_info *shminfo = platform_heap_id_to_shmaddr(hid);
+
+   debug_assert(
+      (platform_shm_heap_handle_valid((platform_heap_handle *)shminfo) == TRUE),
+      "Shared memory heap ID at %p is not a valid shared memory handle.",
+      hid);
+
+   platform_assert(
+      ((((uint64)shminfo->shm_next) % PLATFORM_CACHELINE_SIZE) == 0),
+      "[%s:%d] Next free-addr is not aligned: "
+      "shm_next=%p, shm_total_bytes=%lu, shm_used_bytes=%lu"
+      ", shm_free_bytes=%lu",
+      file,
+      lineno,
+      shminfo->shm_next,
+      shminfo->shm_total_bytes,
+      shminfo->shm_used_bytes,
+      shminfo->shm_free_bytes);
+
+   void *retptr = NULL;
+
+   // See if we can satisfy requests for large memory fragments from a cached
+   // list of used/free fragments that are tracked separately.
+   if ((size >= SHM_LARGE_FRAG_SIZE)
+       && ((retptr = platform_shm_find_free(
```

Comment:
Control farms-off here to handle large free-fragment recycling scheme. Look into this minion later on.